File: CMPS3500_Minischeme_Group08/procedural/let.cpp
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga  /
  / ASGT: CHECKPOINT 2                                   /
  / ORGN: CSUB - CMPS 3500                               /
  / FILE: let.cpp                                        /
  / DATE: 04/11/2026                                     /
*/

#include "let.h"
#include "evaluate.h"
#include "scope.h"

static std::vector<std::string> extractLetValue(
    const std::vector<std::string>& expr,
    int& i
)
{
    std::vector<std::string> value_expr;

    if (i >= (int)expr.size())
    {
        return value_expr;
    }

    if (expr[i] != "(")
    {
        value_expr.push_back(expr[i]);
        i++;
        return value_expr;
    }

    int depth = 0;

    while (i < (int)expr.size())
    {
        value_expr.push_back(expr[i]);

        if (expr[i] == "(")
        {
            depth++;
        }
        else if (expr[i] == ")")
        {
            depth--;

            if (depth == 0)
            {
                i++;
                break;
            }
        }

        i++;
    }

    return value_expr;
}

// Handles the built-in let expression
std::string handleLet(const std::vector<std::string>& expr, Scope* scope)
{
    Scope* new_scope = enterScope(scope);

    int i = 1;

    if (i >= (int)expr.size() || expr[i] != "(") return "PARSE_ERROR";
    i++;

    while (i < (int)expr.size() && expr[i] == "(")
    {
        i++;

        if (i >= (int)expr.size())
        {
            exitScope(new_scope);
            return "PARSE_ERROR";
        }

        std::string var = expr[i];
        i++;

        std::vector<std::string> value_expr = extractLetValue(expr, i);

        if (value_expr.empty() || i >= (int)expr.size())
        {
            exitScope(new_scope);
            return "PARSE_ERROR";
        }

        i++;

        std::string value = evaluate(value_expr, scope);
        if (value == "PARSE_ERROR" ||
            value == "UNDECLARED_IDENTIFIER" ||
            value == "WRONG_ARITY" ||
            value == "TYPE_MISMATCH" ||
            value == "DIVISION_BY_ZERO")
        {
            exitScope(new_scope);
            return value;
        }
        addScopeEntry(new_scope, var, value);
    }

    
    if (i >= (int)expr.size() || expr[i] != ")")
    {
        exitScope(new_scope);
        return "PARSE_ERROR";
    }
    i++;

    std::vector<std::string> body;

    // Extract the body of the let expression
    while (i < (int)expr.size())
    {
        body.push_back(expr[i]);
        i++;
    }

    std::string result = evaluate(body, new_scope);

    exitScope(new_scope);

    return result;
}

File: CMPS3500_MiniScheme_Starter_Package/oop/Let.java
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga /
  / ASGT: CHECKPOINT 3                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: Let.java                                       /
  / DATE: 05/01/2026                                    /
*/

import java.util.List;
import java.util.ArrayList;

public class Let extends ExpressionHandler {
    public Let(List<String> expression){
        super(expression, 1);
    }

    public static String handleLet(List<String> expression, Scope scope){
        Let letExpression = new Let(expression);
        Scope newScope = scope.enterScope();

        if(letExpression.i >= expression.size() || !expression.get(letExpression.i).equals("(")){
            return "PARSE_ERROR";
        }

        letExpression.i++;

        while(letExpression.i < expression.size() && expression.get(letExpression.i).equals("(")){
            letExpression.i++;

            if(letExpression.i >= expression.size()){
                newScope.exitScope();
                return "PARSE_ERROR";
            }

            String var = expression.get(letExpression.i);
            letExpression.i++;

            List<String> valueExpression = letExpression.extractPart();

            if(valueExpression.isEmpty() || letExpression.i >= expression.size()){
                newScope.exitScope();
                return "PARSE_ERROR";
            }

            letExpression.i++;

            String value = Evaluate.evaluate(valueExpression, newScope);

            if(getError(value)){
                newScope.exitScope();
                return "PARSE_ERROR";
            }

            newScope.addScopeEntry(var, value);
        }

        if(letExpression.i >= expression.size() || !expression.get(letExpression.i).equals(")")){
                newScope.exitScope();
                return "PARSE_ERROR";
        }

        letExpression.i++;

        List<String> body = new ArrayList<>();

        while(letExpression.i < expression.size()){
            body.add(expression.get(letExpression.i));
            letExpression.i++;
        }

        String result = Evaluate.evaluate(body, newScope);
        newScope.exitScope();
        return result;
    }
}
